src/lib/prov/pkcs11/p11_ecdh.cpp

```diff
@@ -0,0 +1,151 @@
+/*
+* PKCS#11 ECDH
+* (C) 2016 Daniel Neus, Sirrix AG
+* (C) 2016 Philipp Weber, Sirrix AG
+*
+* Botan is released under the Simplified BSD License (see license.txt)
+*/
+
+#include <botan/p11_ecdh.h>
+
+#if defined(BOTAN_HAS_ECDH)
+
+#include <botan/internal/p11_mechanism.h>
+#include <botan/ber_dec.h>
+#include <botan/der_enc.h>
+#include <botan/internal/algo_registry.h>
+#include <botan/internal/pk_utils.h>
+
+#if defined(BOTAN_HAS_SYSTEM_RNG)
+   #include <botan/system_rng.h>
+#else
+   #include <botan/auto_rng.h>
+#endif
+
+namespace Botan {
+
+namespace PKCS11 {
+
+ECDH_PublicKey PKCS11_ECDH_PublicKey::export_key() const
+   {
+   return ECDH_PublicKey(domain(), public_point());
+   }
+
+ECDH_PrivateKey PKCS11_ECDH_PrivateKey::export_key() const
+   {
+
+#if defined(BOTAN_HAS_SYSTEM_RNG)
+   System_RNG rng;
```

Comment:
I'm pretty sure since this rng object is not even called, because `x` is > 0, so a `Null_RNG` would also work. Any use of `Null_RNG` indicates a case where the application knows the rng will not be used but an rng is syntactically required by the API, as is here. (And typically points to an API problem) 

The need to create this RNG which is not even used could be fixed by splitting the ECDH_PrivateKey (and ECC_PrivateKey) constructors into two versions, one taking an rng and the other the specific value, instead of having one constructor taking both values and choosing which to use at runtime. That would be a cleaner interface since as it is every caller must supply a dummy value (either the rng or the BigInt). I will add this to the long list of cleanups...